mlbuffer-exploration.md

```diff
@@ -0,0 +1,395 @@
+  - If we know the usage of `inputMlBuffer` (e.g. that it's read-only by WebNN)
+    then we may be able to eliminate the data copy from the "upload" buffer to
+    the high-GPU-bandwidth buffer in the non-UMA case:
+    ```js
+    const mlBuffer = mlContext.createBuffer({size, usage: MAP_WRITE | INPUT_ONLY});
+    ```
+    This may not make a difference for DirectML, which appears to require [bound
+    resources](https://learn.microsoft.com/en-us/windows/win32/api/directml/nf-directml-idmlbindingtable-bindpersistentresource)
+    to use `D3D12_HEAP_TYPE_DEFAULT`, but it could eliminate a copy on other
+    systems. I'm not familiar enough with other systems to know the answer here!
+  - Combining this with the above techniques brings (as many as) 4 copies down
+    to (as few as) 2:\
+    &emsp;&emsp;Before: *some source &rarr; `ArrayBuffer` &rarr; "upload" buffer
+      &rarr; high-GPU-bandwidth buffer*\
+    &emsp;&emsp;After: *some source &rarr; "upload" buffer*
+
+### Execute an `MLGraph`
+
+```js
+mlContext.dispatch(
+    graph,
+    /*inputs=*/{buffer: inputMlBuffer},
+    /*outputs=*/{buffer: intermediateMlBuffer},
+);
+```
+
+#### How it works:
+
+- Enqueues a request to compute the graph onto some WebNN timeline
+- Execution cannot start until all input and output `MLBuffer`s are available
+- All input and output `MLBuffer`s are unavailable while execution is in
+  progress
+- All work submitted after this `dispatch()` call which relies on an input or
+  output `MLBuffer` will be queued behind this execution
+
+#### Questions:
+
+- This approach is flexible enough to allow for graph execution on all backends.
+  Do we need a separate `compute()` method?
+- Should this method be on the `MLGraph` (related to
+  [#303](https://github.com/webmachinelearning/webnn/issues/303))? Is there a
+  use case not satisfied by the following?
+  ```js
+  graph.dispatch(
+    /*inputs=*/{buffer: inputMlBuffer},
+    /*outputs=*/{buffer: intermediateMlBuffer},
+  );
+  ```
+- Is it valid to pass the same `MLBuffer` as both an input and output of the
+  same `dispatch()` call? e.g.
+  ```js
+  graph.dispatch(
+    /*inputs=*/{buffer: someMlBuffer},
+    /*outputs=*/{buffer: someMlBuffer},
+  );
+  ```
+
+### Read back data from an `MLBuffer`
+
+```js
+const resultBuffer = await outputMlBuffer.mapAsync();
+```
+
+#### How it works:
+
+- After the completion of all currently-enqueued operations that use
+  `outputMlBuffer`, WebNN will copy the contents of `outputMlBuffer` to
+  `resultBuffer`. This is very similar to
+  [`GPUBuffer.mapAsync()`](https://www.w3.org/TR/webgpu/#dom-gpubuffer-mapasync),
+  with a key difference being that WebGPU only allows `mapAsync()` on buffers
+  which have the `MAP_READ` usage flag. In this case, if using a GPU, we may
+  need to create an intermediate "readback" buffer to facilitate the transfer.
+  This may require two copies:\
+  *&emsp;&emsp;high-GPU-bandwidth buffer &rarr; "readback" buffer &rarr;
+  `ArrayBuffer`*\
```

Comment:
For UMA GPU, the implementation may eliminate the "readback" buffer and just need one copy.